src/annotator/integrations/vitalsource.js

```diff
@@ -0,0 +1,208 @@
+import { ListenerCollection } from '../../shared/listener-collection';
+import { HTMLIntegration } from './html';
+
+/**
+ * @typedef {import('../../types/annotator').Anchor} Anchor
+ * @typedef {import('../../types/annotator').Annotator} Annotator
+ * @typedef {import('../../types/annotator').Integration} Integration
+ * @typedef {import('../../types/annotator').Selector} Selector
+ */
+
+/**
+ * Return the custom DOM element that contains the book content iframe.
+ */
+function findBookElement(document_ = document) {
+  return document_.querySelector('mosaic-book');
+}
+
+/**
+ * Return the role of the current frame in the VitalSource Bookshelf reader or
+ * `null` if the frame is not part of Bookshelf.
+ *
+ * @return {'container'|'content'|null} - `container` if this is the parent of
+ *   the content frame, `content` if this is the frame that contains the book
+ *   content or `null` if the document is not part of the Bookshelf reader.
+ */
+export function vitalSourceFrameRole(window_ = window) {
+  if (findBookElement(window_.document)) {
+    return 'container';
+  }
+
+  const parentDoc = window_.frameElement?.ownerDocument;
+  if (parentDoc && findBookElement(parentDoc)) {
+    return 'content';
+  }
+
+  return null;
+}
+
+/**
+ * Integration for the container frame in VitalSource's Bookshelf ebook reader.
+ *
+ * This frame cannot be annotated directly. This integration serves only to
+ * load the client into the frame that contains the book content.
+ *
+ * @implements {Integration}
+ */
+export class VitalSourceContainerIntegration {
+  /**
+   * @param {Annotator} annotator
+   */
+  constructor(annotator) {
+    const bookElement = findBookElement();
+    if (!bookElement) {
+      throw new Error('Book container element not found');
+    }
+
+    const shadowRoot = /** @type {ShadowRoot} */ (bookElement.shadowRoot);
+    const injectClientIntoContentFrame = () => {
+      const frame = shadowRoot.querySelector('iframe');
+      if (frame) {
+        annotator.injectClient(frame);
+      }
+    };
+    this._frameObserver = new MutationObserver(injectClientIntoContentFrame);
+    this._frameObserver.observe(shadowRoot, { childList: true, subtree: true });
```

Comment:
This will invoke the callback on any DOM mutation; as we discussed verbally, it might be good to debounce this or otherwise tame it. In any case, either the callback or the `injectClient` method will need to be "intelligent" about when to actually inject the client...